src/System.Text.Primitives/System/Text/Parsers/PrimitiveParser_Guid.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Buffers
+{
+    public static partial class Parsers
+    {
+        public static partial class Utf8
+        {
+            public static bool TryParseGuid(ReadOnlySpan<byte> text, out Guid value, out int bytesConsumed, ParsedFormat format = default(ParsedFormat))
+            {
+                if (format.IsDefault) format = 'D';
+                switch (format.Symbol)
+                {
+                    case 'D': return TryParseGuidCore(text, false, ' ',  ' ', out value, out bytesConsumed);
+                    case 'B': return TryParseGuidCore(text, true, '{', '}', out value, out bytesConsumed);
+                    case 'P': return TryParseGuidCore(text, true, '(', ')', out value, out bytesConsumed);
+                    case 'N': return TryParseGuidN(text, out value, out bytesConsumed);
+                    default: throw new FormatException();
+                }
+            }
+
+            private static bool TryParseGuidN(ReadOnlySpan<byte> text, out Guid value, out int bytesConsumed)
+            {
+                bytesConsumed = 0;
+                if (text.Length < 32) return false;
+
+                int justConsumed;
+                if (!Hex.TryParseUInt32(text.Slice(0, 8), out uint i1, out justConsumed) || justConsumed != 8) return false; // 8 digits
+                if (!Hex.TryParseUInt16(text.Slice(8, 4), out ushort i2, out justConsumed) || justConsumed != 4) return false; // next 4 digfits
+                if (!Hex.TryParseUInt16(text.Slice(12, 4), out ushort i3, out justConsumed) || justConsumed != 4) return false; // next 4 digits
+                if (!Hex.TryParseUInt16(text.Slice(16, 4), out ushort i4, out justConsumed) || justConsumed != 4) return false; // next 4 digits
+                if (!Hex.TryParseUInt64(text.Slice(20), out ulong i5, out justConsumed) || justConsumed != 12) return false; // next 12 digits
+
+                bytesConsumed = 32;
+                value = new Guid((int)i1, (short)i2, (short)i3, (byte)(i4 >> 8), (byte)i4,
+                    (byte)(i5 >> 40), (byte)(i5 >> 32), (byte)(i5 >> 24), (byte)(i5 >> 16), (byte)(i5 >> 8), (byte)i5);
+                return true;
+            }
+
+            // {8-4-4-4-12}, where number is the number of hex digits, and {/} are ends.
+            private static bool TryParseGuidCore(ReadOnlySpan<byte> text, bool ends, char begin, char end, out Guid value, out int bytesConsumed)
+            {
+                bytesConsumed = 0;
+                var expectedCodingUnits = 36 + (ends ? 2 : 0); // 32 hex digits + 4 delimiters + 2 optional ends
+
+                if (text.Length < expectedCodingUnits) return false;
+
+                if (ends) 
+                {
+                    if (text[0] != begin) return false;
```

Comment:
Can we combine the last character check with the first character check, and remove this line `if ends && text[justConsumed] != end) return false;` (https://github.com/dotnet/corefxlab/pull/1768/files#diff-f6ac7ef6f085edf5be4c1268e0e0f6ccR82)
```C#
if (text[0] != begin || text[expectedCodingUnits - 1] != end) return false
```